translator/trace/otlp_to_oc.go

```
@@ -0,0 +1,400 @@
+}
+
+func attributeValueToString(attr *otlpcommon.AttributeKeyValue) string {
+	switch attr.Type {
+	case otlpcommon.AttributeKeyValue_STRING:
+		return attr.StringValue
+	case otlpcommon.AttributeKeyValue_BOOL:
+		return strconv.FormatBool(attr.BoolValue)
+	case otlpcommon.AttributeKeyValue_DOUBLE:
+		return strconv.FormatFloat(attr.DoubleValue, 'f', -1, 64)
+	case otlpcommon.AttributeKeyValue_INT:
+		return strconv.FormatInt(attr.IntValue, 10)
+	default:
+		return fmt.Sprintf("<Unknown OpenTelemetry attribute value type %q>", attr.Type)
+	}
+}
+
+func attributeValueToOC(attr *otlpcommon.AttributeKeyValue) *octrace.AttributeValue {
+	a := &octrace.AttributeValue{}
+
+	switch attr.Type {
+	case otlpcommon.AttributeKeyValue_STRING:
+		a.Value = &octrace.AttributeValue_StringValue{
+			StringValue: &octrace.TruncatableString{
+				Value: attr.StringValue,
+			},
+		}
+	case otlpcommon.AttributeKeyValue_BOOL:
+		a.Value = &octrace.AttributeValue_BoolValue{
+			BoolValue: attr.BoolValue,
+		}
+	case otlpcommon.AttributeKeyValue_DOUBLE:
+		a.Value = &octrace.AttributeValue_DoubleValue{
+			DoubleValue: attr.DoubleValue,
+		}
+	case otlpcommon.AttributeKeyValue_INT:
+		a.Value = &octrace.AttributeValue_IntValue{
+			IntValue: attr.IntValue,
+		}
+	default:
+		a.Value = &octrace.AttributeValue_StringValue{
+			StringValue: &octrace.TruncatableString{
+				Value: fmt.Sprintf("<Unknown OpenTelemetry attribute value type %q>", attr.Type),
+			},
+		}
+	}
+
+	return a
+}
+
+// OTLP follows the W3C format, e.g. "vendorname1=opaqueValue1,vendorname2=opaqueValue2"
+// See https://w3c.github.io/trace-context/#combined-header-value
+func tracestateToOC(tracestate string) *octrace.Span_Tracestate {
+	if tracestate == "" {
+		return nil
+	}
+
+	// key-value pairs in the "key1=value1" format
+	pairs := strings.Split(tracestate, ",")
+
+	entries := make([]*octrace.Span_Tracestate_Entry, 0, len(pairs))
+	for _, pair := range pairs {
+		kv := strings.SplitN(pair, "=", 2)
+		key := kv[0]
```

Can SplitN return an empty slice?

---

Yes. Shall I silently skip the entry in that case?

---

I think yes, it should be skipped.